src/credix_lp/depository.ts

```diff
@@ -0,0 +1,286 @@
+import { BorshAccountsCoder, Wallet } from '@project-serum/anchor';
+import { ConfirmOptions, Connection, PublicKey, Signer } from '@solana/web3.js';
+import { IDL } from '../idl';
+import { CredixLpDepositoryAccount } from '../interfaces';
+import { Token, TOKEN_PROGRAM_ID } from '@solana/spl-token';
+import { findATAAddrSync } from '../utils';
+import { AnchorProvider } from '@project-serum/anchor';
+import { IDL as credixIDL, Credix as CredixIDL } from './credixIdl';
+import { Program } from '@project-serum/anchor';
+
+const CREDIX_LP_DEPOSITORY_NAMESPACE = 'CREDIX_LP_DEPOSITORY';
+
+const CREDIX_LP_INTERNAL_CREDIX_MARKETPLACE_NAMESPACE = 'credix-marketplace';
+const CREDIX_LP_INTERNAL_CREDIX_PASS_NAMESPACE = 'credix-pass';
+const CREDIX_LP_INTERNAL_PROGRAM_STATE_NAMESPACE = 'program-state';
+
+export class CredixLpDepository {
+  public constructor(
+    public readonly pda: PublicKey,
+    public readonly collateralMint: PublicKey,
+    public readonly collateralDecimals: number,
+    public readonly collateralSymbol: string,
+    public readonly depositoryCollateral: PublicKey,
+    public readonly depositoryShares: PublicKey,
+    public readonly credixProgramState: PublicKey,
+    public readonly credixGlobalMarketState: PublicKey,
+    public readonly credixSigningAuthority: PublicKey,
+    public readonly credixLiquidityCollateral: PublicKey,
+    public readonly credixSharesMint: PublicKey,
+    public readonly credixPass: PublicKey,
+    public readonly credixTreasuryCollateral: PublicKey,
+    public readonly credixMultisigKey: PublicKey,
+    public readonly credixMultisigCollateral: PublicKey,
+    public readonly credixProgramId: PublicKey
+  ) {}
+
+  public static async initialize({
+    connection,
+    uxdProgramId,
+    collateralMint,
+    collateralSymbol,
+    credixProgramId,
+  }: {
+    connection: Connection;
+    uxdProgramId: PublicKey;
+    collateralMint: PublicKey;
+    collateralSymbol: string;
+    credixProgramId: PublicKey;
+  }): Promise<CredixLpDepository> {
+    // First we need the credix global market state address
+    const credixGlobalMarketState = await this.findCredixGlobalMarketState(
+      credixProgramId
+    );
+    // Then the credix program state address
+    const credixProgramState = await this.findCredixProgramState(
+      credixProgramId
+    );
```

Comment:
Can it be run asynchonously to make it faster?

```suggestion
    const [
      credixGlobalMarketState,
      credixProgramState,
    ] = await Promise.all([
         // First we need the credix global market state address
         this.findCredixGlobalMarketState(
            credixProgramId
         ),
         // Then the credix program state address
         this.findCredixProgramState(
            credixProgramId
         ),
    ]);
```

Reply:
Yep it could be run through a promise for sure, but that Promise.all syntax is much more complex to read and prone to error.

From what I can understand, this code is never ran in a tight loop so the performance difference here sounds very inconsequential on any PC/mobile. And this type of syntax has a very real readability cost.

I explicitely was trying to avoid syntax like this because to me in this case this belong in the premature optimization category. Some detailled explainatio here:
https://softwareengineering.stackexchange.com/questions/80084/is-premature-optimization-really-the-root-of-all-evil

Reply:
Also opiniated.

In this case the benefits of `Promise.all` may be small considering the underlying operation (`findProgramAddress`). So ok.

>  this code is never ran in a tight loop so the performance difference here sounds very inconsequential on any PC/mobile

This code is the part of the credix depository initialization, that is called in `initServices` function. As I understand it, while credix/mercurial services are loading, the app isn't started. In this case it's ok because it's `findProgramAddress` calls. If it were API calls for example, would delay app initialization for 100s ms which would be problematic.

```
export default async function init(config: ISolanaConfiguration) {
  // async init functions that can safely be ran in parallel with the rest
  const statusP = initSolanaStatus();

  const clusterRPCEndpoint = await initClusterRPCEndpoint(config);
  const connection = initConnection(clusterRPCEndpoint);

  const servicesP = initServices(config, connection);

  const [status, services] = await Promise.all([statusP, servicesP]);

  return { config, clusterRPCEndpoint, connection, status, services };
}
```

Reply:
Ok that make sense, so if this is a bottle neck for the "first rendering" this might be something worth optimizing.

In that case however we could optimize in a different way that is much more readable: by delaying the awaits rather than using Promise.All()

```tsx
const aValue = await doMyThingA();
const bValue = await doMyThingB();
doMyThingyWithA(aValue);
doMyThingyWithB(bValue);
```

for example this is a much more readable pattern:

```tsx
const aValuePromise = doMyThingA();
const bValuePromise = doMyThingB();
doMyThingyWithA(await aValuePromise);
doMyThingyWithB(await bValuePromise);
```

rather than:

```tsx
const [aValue, bValue] = Promise.All([
    doMyThingA(),
    doMyThingB()
]);
doMyThingyWithA(aValue);
doMyThingyWithB(bValue);
```

As you can probably tell im no big fan of Promise.All, but that's because it goes against keeping the compiler in control of the syntax (ordering can easily be a problem for large codebases)

Reply:
I will change the credix functions, but I would recommend at all times against using Promise.All() or any similar order-based functions

Reply:
I don't get your hate of `Promise.all` tbh, it's an useful tool, and quite readeable in the contrary.

Using:

```
const aValuePromise = doMyThingA();
const bValuePromise = doMyThingB();
doMyThingyWithA(await aValuePromise);
doMyThingyWithB(await bValuePromise);
```

also works, but keep in mind the potential disadvantages:

- You have to create temporary variables to store `Promise` objects that have to be resolved manually. You may forget to resolve them and as a result end up with unresolved Promises. 

Error handling:
-`aValuePromise` are decorallated. If `bValuePromise` fails, you will still execute `doMyThingyWithA`. In a lot of usecases, if `doMyThingA` or `doMyThingB` fails, you want to exit the function and do nothing.
- If `await aValuePromise` throw an error, then, you end-up with `bValuePromise` to never be resolved. Most cases, the throw will makes a change of scope to happens leading to the `bValuePromise` and then the Promise object  itself (in _pending_ state) to be garbage collected.

Reply:
I'm ok with both `Promise.all` and manual Promise resolution. It depends of your needs and the case you are facing

Reply:
If you have 1 API call to make per account in a given array, `Promise.all` is great option for example:

```
   const accountInfo = await Promise.all(accounts.map(account => account.publicKey));
```

Reply:
Not sure I get:

> but that's because it goes against keeping the compiler in control of the syntax/control flow (ordering can easily be a problem for large codebases)

`Promise.all` wraps given Promises into a single `Promise`. Not sure I follow you there.

Reply:
> Not sure I get:
> 
> > but that's because it goes against keeping the compiler in control of the syntax/control flow (ordering can easily be a problem for large codebases)
> 
> `Promise.all` wraps given Promises into a `Promise`. Not sure I follow you there.

My point was that the typing/naming information mostly lost when passing through that function (names are visually enforced by the developer and types are lost except when having the same exact type)




> If you have 1 API call to make per account in a given array, `Promise.all` is great option for example:
> 
> ```
>    const accountInfo = await Promise.all(accounts.map(account => account.publicKey));
> ```

Yeah thats definitly a good usecase for it of course, but thats quite particular case of: everything is of the same type and the input is an array.




> I don't get your hate of `Promise.all` tbh, it's an useful tool, and quite readeable in the contrary.

I think my main point comes from the error-prone-ness of it, when your input is not an array (instead its a bunch of variables) It's much better to construct the "dependency tree of promise resolution" through the code by using async/await rather than forcing the order of execution manually using a Promise.all.

For example: in the above example: `doMyThingyWithA` will already be running even before `bValuePromise` has resolved, which would not be the case with a Promise.all().

Overall my point is that 90% of the time (but not 100%) we are better off avoiding Promise.all for purely syntax purposes